src/isa/riscv64/system/priv.c

```diff
@@ -1335,14 +1531,50 @@ static word_t priv_instr(uint32_t op, const rtlreg_t *src) {
 #ifdef CONFIG_RVH
       cpu.v = (mstatus->mpp == MODE_M ? 0 : mstatus->mpv);
       mstatus->mpv = 0;
+#ifdef CONFIG_RV_SSDBLTRP
+      vsstatus->sdt = (cpu.mode == MODE_U && cpu.v == 1 ? 0 : vsstatus->sdt);
+#endif // CONFIG_RV_SSDBLTRP
       set_sys_state_flag(SYS_STATE_FLUSH_TCACHE);
 #endif // CONFIG_RVH
       if (mstatus->mpp != MODE_M) { mstatus->mprv = 0; }
       mstatus->mpp = MODE_U;
+#ifdef CONFIG_RV_SMDBLTRP
+      mstatus->mdt = 0;
+#endif // CONFIG_RV_SMDBLTRP
+#ifdef CONFIG_RV_SSDBLTRP
+      mstatus->sdt = (cpu.mode == MODE_M ? mstatus->sdt : 0);
+#endif // CONFIG_RV_SSDBLTRP
       update_mmu_state();
       Loge("Executing mret to 0x%lx", mepc->val);
       return mepc->val;
       break;
+#ifdef CONFIG_RV_SMRNMI
+    case 0x702: // mnret
+      if (cpu.mode < MODE_M) {
+        longjmp_exception(EX_II);
+      }
+      if (mnstatus->mnpp != MODE_M) { mstatus->mprv = 0; }
+      cpu.mode = mnstatus->mnpp;
+#ifdef CONFIG_RVH
+      cpu.v    = (mnstatus->mnpp == MODE_M ? 0 : mnstatus->mnpv);
+      mnstatus->mnpv = 0;
+      // clear vsstatus.SDT when return to VU
+      vsstatus->sdt = (cpu.mode == MODE_U && cpu.v == 1 ? 0 : vsstatus->sdt);
+      set_sys_state_flag(SYS_STATE_FLUSH_TCACHE);
+#endif // config_RVH
+      mnstatus->mnpp = MODE_U;
+      mnstatus->nmie = 1;
+
+      // clear MDT when mnret to below M
+      mstatus->mdt = (cpu.mode == MODE_M ? mstatus->mdt: 0);
+      mstatus->sdt = (cpu.mode == MODE_M ? mstatus->sdt: 0);
```

Comment:
mnstatus->mnpp is better than cpu.mode, since cpu.mode is just assigned in this function.
When using state machine design pattern, all state changes should be assigned with current state(mnstatus->mnpp) or the condition of state transfer directly, not the next state of other field(here cpu.mode) which is just assigned to.